Node crash due to Postgres

<!-- Describe the issue -->

Node failure due to `Cannot check database lease`. Multiple times, didn't notice this issue with default Postgres configs.

<!-- What behaviour did you expect? -->

Maybe it would be good for node to not crash and just reject some operations with channels?

<!-- How reliably can you reproduce the issue, what are the steps to do so? -->

It repeats regularly, recently it lead to channel force-close.

<!-- What version of Eclair are you using, where did you get it (website, self-compiled, etc)? -->

Self-compiled. "0.5.1-98bb7be"

<!-- What environment are you observing the error on (OS, JDK version)? -->

Linux Debian 10. OpenJDK 64-Bit Server VM (build 11.0.9.1+1-post-Debian-1deb10u2, mixed mode, sharing)

<!-- Any extra information that might be useful in the debugging process. -->

I had config
```
eclair.db.postgres.lease {
    interval = 60 seconds        // Default: 5 minutes
    renew-interval =  30 seconds // Default: 1 minute
}
```
I made lease times shorter for faster node restarts. What is suggested time intervals?

<!-- This is normally the contents of a `eclair.log` or `eclair.conf` file. Raw text or a link to a pastebin type site are preferred. -->

```
2021-05-05 16:36:41,615 ERROR f.a.eclair.db.Databases - fatal error: Cannot obtain lock on the database.
fr.acinq.eclair.db.pg.PgUtils$LockException: Cannot check database lease
	at fr.acinq.eclair.db.pg.PgUtils$.fr$acinq$eclair$db$pg$PgUtils$$checkDatabaseLease(PgUtils.scala:211)
	at fr.acinq.eclair.db.pg.PgUtils$LeaseLock.$anonfun$withLock$1(PgUtils.scala:94)
	at fr.acinq.eclair.db.pg.PgUtils$.inTransaction(PgUtils.scala:106)
	at fr.acinq.eclair.db.pg.PgUtils$.$anonfun$inTransaction$1(PgUtils.scala:121)
	at fr.acinq.eclair.db.jdbc.JdbcUtils.withConnection(JdbcUtils.scala:35)
	at fr.acinq.eclair.db.jdbc.JdbcUtils.withConnection$(JdbcUtils.scala:32)
	at fr.acinq.eclair.db.pg.PgUtils$.withConnection(PgUtils.scala:30)
	at fr.acinq.eclair.db.pg.PgUtils$.inTransaction(PgUtils.scala:120)
	at fr.acinq.eclair.db.pg.PgUtils$LeaseLock.withLock(PgUtils.scala:92)
	at fr.acinq.eclair.db.pg.PgChannelsDb.$anonfun$addOrUpdateChannel$1(PgChannelsDb.scala:53)
	at fr.acinq.eclair.db.Monitoring$Metrics$.$anonfun$withMetrics$1(Monitoring.scala:33)
	at fr.acinq.eclair.KamonExt$.time(PimpKamon.scala:28)
	at fr.acinq.eclair.db.Monitoring$Metrics$.withMetrics(Monitoring.scala:31)
	at fr.acinq.eclair.db.pg.PgChannelsDb.addOrUpdateChannel(PgChannelsDb.scala:53)
	at fr.acinq.eclair.channel.Channel$MyState.storing(Channel.scala:2434)
	at fr.acinq.eclair.channel.Channel$$anonfun$9.applyOrElse(Channel.scala:774)
	at fr.acinq.eclair.channel.Channel$$anonfun$9.applyOrElse(Channel.scala:656)
	at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:35)
	at fr.acinq.eclair.channel.Channel$$anonfun$handleExceptions$1.applyOrElse(Channel.scala:2415)
	at fr.acinq.eclair.channel.Channel$$anonfun$handleExceptions$1.applyOrElse(Channel.scala:2412)
	at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:35)
	at akka.actor.FSM.processEvent(FSM.scala:804)
	at akka.actor.FSM.processEvent$(FSM.scala:801)
	at fr.acinq.eclair.channel.Channel.processEvent(Channel.scala:103)
Caused by: org.postgresql.util.PSQLException: ERROR: could not serialize access due to read/write dependencies among transactions
  Detail: Reason code: Canceled on identification as a pivot, during conflict out checking.
  Hint: The transaction might succeed if retried.
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2533)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2268)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:313)
	at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:448)
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:369)
	at org.postgresql.jdbc.PgStatement.executeWithFlags(PgStatement.java:310)
	at org.postgresql.jdbc.PgStatement.executeCachedSql(PgStatement.java:296)
	at org.postgresql.jdbc.PgStatement.executeWithFlags(PgStatement.java:273)
	at org.postgresql.jdbc.PgStatement.executeQuery(PgStatement.java:226)
	at com.zaxxer.hikari.pool.ProxyStatement.executeQuery(ProxyStatement.java:111)
	at com.zaxxer.hikari.pool.HikariProxyStatement.executeQuery(HikariProxyStatement.java)
	at fr.acinq.eclair.db.pg.PgUtils$.$anonfun$getCurrentLease$1(PgUtils.scala:221)
	at fr.acinq.eclair.db.jdbc.JdbcUtils.using(JdbcUtils.scala:50)
	at fr.acinq.eclair.db.jdbc.JdbcUtils.using$(JdbcUtils.scala:46)
	at fr.acinq.eclair.db.pg.PgUtils$.using(PgUtils.scala:30)
	at fr.acinq.eclair.db.pg.PgUtils$.getCurrentLease(PgUtils.scala:220)
	at fr.acinq.eclair.db.pg.PgUtils$.$anonfun$checkDatabaseLease$1(PgUtils.scala:197)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
	at scala.util.Try$.apply(Try.scala:210)
	at fr.acinq.eclair.db.pg.PgUtils$.fr$acinq$eclair$db$pg$PgUtils$$checkDatabaseLease(PgUtils.scala:197)
	at fr.acinq.eclair.db.pg.PgUtils$LeaseLock.$anonfun$withLock$1(PgUtils.scala:94)
	at fr.acinq.eclair.db.pg.PgUtils$.inTransaction(PgUtils.scala:106)
	at fr.acinq.eclair.db.pg.PgUtils$.$anonfun$inTransaction$1(PgUtils.scala:121)
	at fr.acinq.eclair.db.jdbc.JdbcUtils.withConnection(JdbcUtils.scala:35)
```

@rorp

--- a/eclair-core/src/main/scala/fr/acinq/eclair/db/pg/PgChannelsDb.scala
+++ b/eclair-core/src/main/scala/fr/acinq/eclair/db/pg/PgChannelsDb.scala
@@ -16,6 +16,7 @@
 
 package fr.acinq.eclair.db.pg
 
+import com.zaxxer.hikari.util.IsolationLevel
 import fr.acinq.bitcoin.ByteVector32
 import fr.acinq.eclair.CltvExpiry
 import fr.acinq.eclair.channel.HasCommitments
@@ -102,7 +103,7 @@ class PgChannelsDb(implicit ds: DataSource, lock: PgLock) extends ChannelsDb wit
    * Helper method to factor updating timestamp columns
    */
   private def updateChannelMetaTimestampColumn(channelId: ByteVector32, columnName: String): Unit = {
-    inTransaction { pg =>
+    inTransaction(IsolationLevel.TRANSACTION_READ_UNCOMMITTED) { pg =>
       using(pg.prepareStatement(s"UPDATE local_channels SET $columnName=? WHERE channel_id=?")) { statement =>
         statement.setTimestamp(1, Timestamp.from(Instant.now()))
         statement.setString(2, channelId.toHex)

--- a/eclair-core/src/main/scala/fr/acinq/eclair/db/pg/PgUtils.scala
+++ b/eclair-core/src/main/scala/fr/acinq/eclair/db/pg/PgUtils.scala
@@ -16,6 +16,7 @@
 
 package fr.acinq.eclair.db.pg
 
+import com.zaxxer.hikari.util.IsolationLevel
 import fr.acinq.eclair.db.Monitoring.Metrics._
 import fr.acinq.eclair.db.Monitoring.Tags
 import fr.acinq.eclair.db.jdbc.JdbcUtils
@@ -23,7 +24,7 @@ import fr.acinq.eclair.db.pg.PgUtils.PgLock.LockFailureHandler.LockException
 import grizzled.slf4j.Logging
 import org.postgresql.util.{PGInterval, PSQLException}
 
-import java.sql.{Connection, Statement, Timestamp}
+import java.sql.{Connection, Timestamp}
 import java.util.UUID
 import javax.sql.DataSource
 import scala.concurrent.duration._
@@ -242,11 +243,14 @@ object PgUtils extends JdbcUtils {
 
   }
 
-  def inTransaction[T](connection: Connection)(f: Connection => T): T = {
-    val autoCommit = connection.getAutoCommit
+  /**
+   * @param isolationLevel Be careful when changing the default value
+   */
+  private def inTransactionInternal[T](isolationLevel: IsolationLevel)(connection: Connection)(f: Connection => T): T = {
+    val previousAutoCommit = connection.getAutoCommit
     connection.setAutoCommit(false)
-    val isolationLevel = connection.getTransactionIsolation
-    connection.setTransactionIsolation(Connection.TRANSACTION_SERIALIZABLE)
+    val previousIsolationLevel = connection.getTransactionIsolation
+    connection.setTransactionIsolation(isolationLevel.getLevelId)
     try {
       val res = f(connection)
       connection.commit()
@@ -256,14 +260,20 @@ object PgUtils extends JdbcUtils {
         connection.rollback()
         throw ex
     } finally {
-      connection.setAutoCommit(autoCommit)
-      connection.setTransactionIsolation(isolationLevel)
+      connection.setAutoCommit(previousAutoCommit)
+      connection.setTransactionIsolation(previousIsolationLevel)
     }
   }
 
   def inTransaction[T](f: Connection => T)(implicit dataSource: DataSource): T = {
     withConnection { connection =>
-      inTransaction(connection)(f)
+      inTransactionInternal(IsolationLevel.TRANSACTION_SERIALIZABLE)(connection)(f)
+    }
+  }
+
+  def inTransaction[T](isolationLevel: IsolationLevel)(f: Connection => T)(implicit dataSource: DataSource): T = {
+    withConnection { connection =>
+      inTransactionInternal(isolationLevel)(connection)(f)
     }
   }
 

--- a/eclair-core/src/test/scala/fr/acinq/eclair/db/ChannelsDbSpec.scala
+++ b/eclair-core/src/test/scala/fr/acinq/eclair/db/ChannelsDbSpec.scala
@@ -98,8 +98,11 @@ class ChannelsDbSpec extends AnyFunSuite {
       val db = dbs.channels
       implicit val ec: ExecutionContextExecutor = ExecutionContext.fromExecutor(Executors.newFixedThreadPool(8))
       val channel = ChannelCodecsSpec.normal
-      val futures = for (_ <- 0 until 10000) yield {
-        Future(db.addOrUpdateChannel(channel.modify(_.commitments.channelId).setTo(randomBytes32)))
+      val channelIds = (for(_ <- 0 until 10) yield randomBytes32).toList
+      val futures = for (i <- 0 until 10000) yield {
+        val channelId = channelIds(i % channelIds.size)
+        Future(db.addOrUpdateChannel(channel.modify(_.commitments.channelId).setTo(channelId)))
+        Future(db.updateChannelMeta(channelId, ChannelEvent.EventType.PaymentSent))
       }
       val res = Future.sequence(futures)
       Await.result(res, 60 seconds)

--- a/eclair-core/src/test/scala/fr/acinq/eclair/db/ChannelsDbSpec.scala
+++ b/eclair-core/src/test/scala/fr/acinq/eclair/db/ChannelsDbSpec.scala
@@ -98,7 +98,7 @@ class ChannelsDbSpec extends AnyFunSuite {
       val db = dbs.channels
       implicit val ec: ExecutionContextExecutor = ExecutionContext.fromExecutor(Executors.newFixedThreadPool(8))
       val channel = ChannelCodecsSpec.normal
-      val channelIds = (for(_ <- 0 until 10) yield randomBytes32).toList
+      val channelIds = (0 until 10).map(_ => randomBytes32).toList
       val futures = for (i <- 0 until 10000) yield {
         val channelId = channelIds(i % channelIds.size)
         Future(db.addOrUpdateChannel(channel.modify(_.commitments.channelId).setTo(channelId)))

